snapshotter/demux/cache/evict.go

```diff
@@ -0,0 +1,67 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package cache
+
+import (
+	"context"
+	"net"
+	"time"
+)
+
+// EvictionPolicy defines the interface for enforcing a cache eviction policy.
+type EvictionPolicy interface {
+	Enforce(key string)
+}
+
+type evictionPolicy struct {
+	evict chan string
+	stop  chan struct{}
+}
+
+// EvictOnConnectionFailurePolicy defines an eviction policy where entries are evicted
+// from cache after a failed connection attempt occurs.
+type EvictOnConnectionFailurePolicy struct {
+	evictionPolicy
+
+	dial      func(context.Context, string) (net.Conn, error)
+	frequency time.Duration
+}
+
+// NewEvictOnConnectionFailurePolicy creates a new instance with
+func NewEvictOnConnectionFailurePolicy(evictChan chan string, stopCondition chan struct{}, dial func(context.Context, string) (net.Conn, error), frequency time.Duration) EvictionPolicy {
+	return &EvictOnConnectionFailurePolicy{evictionPolicy: evictionPolicy{evict: evictChan, stop: stopCondition}, dial: dial, frequency: frequency}
+}
+
+// Enforce launches a go routine which periodically attempts to dial the cached entry
+// using the provided dial function.
+//
+// On connection failure, the entry will be evicted from cache.
+func (p *EvictOnConnectionFailurePolicy) Enforce(key string) {
+	go func(dial func(context.Context, string) (net.Conn, error)) {
+		ticker := time.NewTicker(p.frequency)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				if _, err := dial(context.Background(), key); err != nil {
+					p.evict <- key
+					return
+				}
+			case <-p.stop:
```

Comment:
On stop, should we also be attempting to evict? Thinking of a scenario where we stop enforcing the policy, but leave an item in the cache, which would leak metrics proxy.

Also just for my understanding: when is `p.stop` signalled? On program termination? The channel is closed on `RemoteSnapshotterCache.Close` but I didn't see it written to anywhere

Reply:
1. Leaving entries on the cache is okay in this case. An eviction policy is applied to a cache which each entry having a background eviction enforcer in the form of this goroutine. This goroutine is only stopped by `RemoteSnapshotterCache.Close` as you noted and then which it precedes to empty the cache.
2. So we just need to close the channel. All receives on a closed channel will result in zero values, in this case an empty struct, and we know it is safe to stop.